Apps/Src/CommandLine.c

```diff
@@ -0,0 +1,357 @@
+
+// Represents a command that the command line understands
+struct Command {
+	const char *name;
+	bool (*action)(void);
+};
+
+static bool cmd_help(void);
+
+static bool cmd_CANbus_Send(void);
+
+static bool cmd_CANbus_Read(void);
+
+static bool cmd_Contactors_Get(void);
+
+static bool cmd_Contactors_Set(void);
+
+static bool cmd_Minion_Read_Input(void);
+
+static bool cmd_Minion_Write_Output(void);
+
+static bool cmd_Pedals_Read(void);
+
+
+const struct Command cmdline_commands[] = {
+	{.name = "help", .action = cmd_help},
+	{.name = "CANbus_Send", .action = cmd_CANbus_Send},
+	{.name = "CANbus_Read", .action = cmd_CANbus_Read},
+	{.name = "Contactors_Get", .action = cmd_Contactors_Get},
+	{.name = "Contactors_Set", .action = cmd_Contactors_Set},
+	{.name = "Minion_Read_Input", .action = cmd_Minion_Read_Input},
+	{.name = "Minion_Write_Output", .action = cmd_Minion_Write_Output},
+	{.name = "Pedals_Read", .action = cmd_Pedals_Read},
+	{.name = NULL, .action = NULL}
+};
+
+static char input[MAX_BUFFER_SIZE];
+char *save; // Save pointer for strtok_r
+
+char *help = {
+	"LHRS Controls Command Line:\n\r"
+	"	For help, enter [help]\n\r"
+	"	Format is: cmd [param, ...]\n\r"
+	"	Commands and their params are as follows:\n\r"
+	"	CANbus_Send (non)blocking motor/car 'string' - Sends a CAN\n\r"
+	"message with the string data as is on the determined line\n\r"
+	"	CANbus_Read (non)blocking motor/car - Reads a CAN message\n\r"
+	"on the detemined line\n\r"
+	"	Contactors_Get array_c/array_p/motor_c - Gets the status of\n\r"
+	"determined contactor\n\r"
+	"	Contactors_Set array_c/array_p/motor_c on/off (non)blocking -\n\r"
+	"Sets the determined contactor\n\r"
+	"	Contactors_Enable  array_c/array_p/motor_c - Enables the determined\n\r"
+	"contactor\n\r"
+	"	Contactors_Disable  array_c/array_p/motor_c - Disables the determined\n\r"
+	"contactor\n\r"
+	"	Minion_Read_Input 'input' - Reads the current status of the input\n\r"
+	"	Minion_Write_Output `output` on/off - Sets the current state of the output\n\r"
+	"	Pedals_Read accel/brake - Reads the current status of the pedal\n\r"
+};
+
+static inline bool isWhiteSpace(char character){
+	switch (character) {
+		case 0x09:
+		case 0x0A:
+		case 0x0C:
+		case 0x0D:
+		case 0x20: return true;
+		default: return false;
+	}
+}
+
+static bool executeCommand(char *input) {
+	// The first word in the input should be a valid command
+	char *command = strtok_r(input, " ", &save);
+	// Iterate through all valid commands and check if the input matches
+	for (int i=0; cmdline_commands[i].name; i++) {
+		if (!strcmp(command, cmdline_commands[i].name)) {
+			return cmdline_commands[i].action(); // Execute the command
+		}
+	}
+	return false; // Didn't find a valid command!
+}
+
+// *********** Command line (shell) ************
+void Task_CommandLine(void* p_arg) { 
+	OS_ERR err;
+
+	// output welcome/help screen
+	printf(help);
+	
+	while(1){
+		printf("> ");
+		BSP_UART_Read(UART_2, input);
```

Comment:
This can probably only be resolved through testing, but I would make sure that this doesn't block without being an OS-integrated blocking call. Because if it blocks and doesn't hit a scheduling point, you won't context switch out. If it's a non-blocking read ur chillin.

Reply:
The function is blocking in that it just has a while loop waiting for a line to be received, which is signaled by the ISR. Not sure why it would be a problem if it blocks. If it blocks, then it acts like an idle task and does nothing and lets the RTOS context switch whenever it needs to. The only other thing to do would be to force a context switch while waiting to not waste CPU cycles but this is not much of an optimization.

Reply:
Are we sure the idle task doesn't hit a scheduling point? I'd be wary of relying on the RTOS to switch without one, but maybe the priority override will handle it. This could be potentially made more "guaranteed" by using a setup similar to the CAN driver, but if it's shown to work then we're chilling.